### PR TITLE
fix: bug in MASTER loss

### PR DESCRIPTION
**Pull request by charlesmindee**
MASTER model was not trained properly due to a misalignment in the loss function, this is now fixed. More precisely:

- Loss function is updated (right shift for gt sequences)
- `encode_sequences` is  improved to support not only `EOS` but also `SOS` and `PAD` symbols, which are mandatory to train MASTER (`SOS` to shift gt to the right, and `PAD` because we need to mask where `PAD` in the loss but nos `EOS`). It is important to note that default behavior of this function is NOT modified, when pad and sos arguments are left to `None`, the function still pad with `EOS` at the end of each sequence. When `PAD` is provided, sequences are concatenated with `EOS` before adding `PAD` to fill the matrix.
- Updated corresponding unitests.
- When gt are given in the call, but `training=False`, the model is now using the prediction decoder to compute the loss, instead of using gts to feed the decoder. This is a more accurate way to evaluate the model and to be sure that we are training the model 'properly': we can check at evaluation time if it is able to predict _from the images only_ the right labels.

Any feedback is welcome!

### CODE REVIEW

**Comment by fg-mindee** [comment]
For reference, #331 is related to this PR

File: doctr/models/recognition/master/tensorflow.py [hunk 1]
```diff
@@ -298,8 +303,13 @@ def call(
             out['loss'] = self.compute_loss(logits, gt, seq_len)
 
         else:
+            # When not training, we want to compute logits in with the decoder, although
+            # we have access to gts (we need gts to compute the loss, but not in the decoder)
             _, logits = self.decode(encoded)
 
+            if target is not None:
+                out['loss'] = self.compute_loss(logits, gt, seq_len)
+
```
Comment (fg-mindee) [inline doctr/models/recognition/master/tensorflow.py:312]: This needs to be unindented ;)

Comment (fg-mindee) [reply doctr/models/recognition/master/tensorflow.py:312]: And you can remove line 302-303 to avoid duplicate loss computation

File: doctr/models/recognition/master/tensorflow.py [hunk 2]
```diff
@@ -290,6 +292,9 @@ def call(
         if target is not None:
             # Compute target: tensor of gts and sequence lengths
             gt, seq_len = self.compute_target(target)
+
+        if training:
+            assert target is not None
```
[the remark this answers is not in the document]
Comment (fg-mindee) [reply doctr/models/recognition/master/tensorflow.py:297]: Actually start with the shortest option:
```python
if target is None:
```
those two conditional branches need to only compute the logits

**Review by fg-mindee — CHANGES_REQUESTED** [review]
I think there is a small mistake in the implementation and then we can merge!

**Review by fg-mindee — APPROVED** [review]
Looks good to me!
